### PR TITLE
add plugin to count load, stores, integer and floating point instructions

### PR DESCRIPTION
**Pull request by GuillermoCallaghan**

### CODE REVIEW

**Comment by lgeek** [comment]
Thanks for writing this plugin. I've ran into an issue testing it. Prefetch instructions seem to trip it up. For example:

```
Error in Load Store counter 0x7fb7fbe010
(gdb) x/i 0x7fb7fbe010
   0x7fb7fbe010:        prfm    pldl1keep, [x1]
(gdb) x/xw 0x7fb7fbe010
0x7fb7fbe010:   0xf9800020
```

I'm not sure if we should count them as loads, create a new counter for NOPs, or do something else. What do you think?

Also, please squash the commits. Thanks

**Comment by GuillermoCallaghan** [comment]
The ARM Reference Manual states that prefetch instructions are implementation-defined. So I think the best thing to do is to have a macro defining if they are counted or not.

File: plugins/instruction_mix.c [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+  fprintf(stderr, "Thread: %d\n", mambo_get_thread_id(ctx));
+
+  // Prints thread private counters
+  print_counters(counters); // comment this out if not needed
+
+  atomic_increment_u64(&global_counters.integer, counters->integer);
+  atomic_increment_u64(&global_counters.floating, counters->floating);
+  atomic_increment_u64(&global_counters.load, counters->load);
+  atomic_increment_u64(&global_counters.store, counters->store);
+  atomic_increment_u64(&global_counters.branch, counters->branch);
```
Comment (lgeek) [inline plugins/instruction_mix.c]: Please increment the global prefetch counter too.

File: plugins/instruction_mix.c [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+  case A64_DATA_PROC_REG1:
+  case A64_CCMP_CCMN_IMMED:
+  case A64_CCMP_CCMN_REG:
+  case A64_COND_SELECT:
+  case A64_LOGICAL_REG:
+    break;
+  case A64_DATA_PROC_REG2:
+    a64_data_proc_reg2_decode_fields(ctx->code.read_address, &sf, &rm, &opcode, &rn, &rd);
+    if ((opcode == 2) || (opcode == 3)) { // UDIV or SDIV
+      inst_counter = &counters->floating;
```
Comment (lgeek) [inline plugins/instruction_mix.c]: Just noticed this, UDIV and SDIV are integer divide instructions, not FP.

File: plugins/instruction_mix.c [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+  case A64_ADD_SUB_EXT_REG:
+  case A64_ADD_SUB_SHIFT_REG:
+  case A64_ADC_SBC:
+    inst_counter = &counters->integer;
+    break;
+  case A64_DATA_PROC_REG1:
+  case A64_CCMP_CCMN_IMMED:
+  case A64_CCMP_CCMN_REG:
+  case A64_COND_SELECT:
+  case A64_LOGICAL_REG:
```
Comment (lgeek) [inline plugins/instruction_mix.c]: I'd classify all of these up to and including A64_DATA_PROC_REG3 as integer instructions.

File: plugins/instruction_mix.c [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  // Data Processing -- Immediate
+  case A64_ADD_SUB_IMMED:
+  case A64_LOGICAL_IMMED:
+    inst_counter = &counters->integer;
+    break;
+
+  case A64_BFM:
+  case A64_ADR:
+  case A64_EXTR:
+  case A64_MOV_WIDE:
```
Comment (lgeek) [inline plugins/instruction_mix.c]: Also integer.

File: plugins/instruction_mix.c (hunk 3 above)
Comment (GuillermoCallaghan) [reply plugins/instruction_mix.c]: Done

File: plugins/instruction_mix.c (hunk 4 above)
Comment (GuillermoCallaghan) [reply plugins/instruction_mix.c]: Done

**Comment by lgeek** [comment]
Merged in f1bfa598cbd9f4

**Pull request closed without merging**